pkg/extend/extend.go

```diff
@@ -0,0 +1,227 @@
+		FirstLBA   int    `json:"firstlba"`
+		LastLBA    int    `json:"lastlba"`
+		Partitions []struct {
+			Node  string `json:"node"`
+			Start int    `json:"start"`
+			Size  int    `json:"size"`
+			Type  string `json:"type"`
+			UUID  string `json:"uuid"`
+			Name  string `json:"name"`
+		}
+	} `json:"partitionTable"`
+}
+
+func autoextend() error {
+	for _, d := range driveKeys {
+		err := exec.Command("sfdisk", "-d", d).Run()
+		if err != nil {
+			log.Printf("No partition table found on device %s. Skipping.", d)
+			continue
+		}
+		if err := extend(d); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func extend(d string) error {
+	data, err := exec.Command("sfdisk", "-J", d).Output()
+	if err != nil {
+		log.Fatalf("Unable to get drive data for %s from sfdisk: %v", d, err)
+	}
+
+	f := Fdisk{}
+	if err := json.Unmarshal(data, &f); err != nil {
+		return fmt.Errorf("Unable to unmarshal partition table from sfdisk: %v", err)
+	}
+
+	if len(f.PartitionTable.Partitions) > 1 {
+		log.Printf("Disk %s has more than 1 partition. Skipping", d)
+		return nil
+	}
+
+	partition := f.PartitionTable.Partitions[0]
+	if partition.Type != "83" {
+		return fmt.Errorf("Partition 1 on disk %s is not a Linux Partition", d)
+	}
+
+	if partition.Start+partition.Size == f.PartitionTable.LastLBA {
+		log.Printf("No free space on device to extend partition")
+		return nil
+	}
+
+	if err := fsck(partition.Node, false); err != nil {
+		return fmt.Errorf("Initial fsck failed: %v", err)
+	}
+
+	if err := exec.Command("sfdisk", "-q", "--delete", d).Run(); err != nil {
+		return fmt.Errorf("Error erasing partition table: %v", err.Error())
+	}
+
+	createCmd := exec.Command("sfdisk", "-q", d)
+	createCmd.Stdin = strings.NewReader(fmt.Sprintf("%d,,83;", partition.Start))
+	if err := createCmd.Run(); err != nil {
+		return fmt.Errorf("Error creating partition table: %v", err)
+	}
+
+	if err := exec.Command("sfdisk", "-A", d, "1").Run(); err != nil {
+		return fmt.Errorf("Error making %s bootable: %v", d, err)
+	}
+
+	// update status
+	if err := exec.Command("blockdev", "--rereadpt", d).Run(); err != nil {
+		return fmt.Errorf("Error running blockdev: %v", err)
+	}
+
+	exec.Command("mdev", "-s").Run()
+
+	// wait for device
+	var done bool
+	for i := 0; i < timeout; i++ {
+		stat, err := os.Stat(partition.Node)
+		if err == nil {
+			mode := stat.Sys().(*syscall.Stat_t).Mode
+			if (mode & syscall.S_IFMT) == syscall.S_IFBLK {
+				done = true
+				break
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+		exec.Command("mdev", "-s").Run()
+	}
+	if !done {
+		return fmt.Errorf("Error waiting for device %s", partition.Node)
+	}
+	// even after the device appears we still have a race
+	time.Sleep(1 * time.Second)
+
+	// resize2fs fails unless we set force=true here
+	if err := fsck(partition.Node, true); err != nil {
+		return fmt.Errorf("fsck before resize failed: %v", err)
+	}
+
+	if err := exec.Command("resize2fs", partition.Node).Run(); err != nil {
```

Comment:
Can we add a filesystem type argument to this too to match the others, even if we just error out if not `ext4` to start with?